CIPs/0010.md

```diff
@@ -0,0 +1,84 @@
+
+- The signing of the off-chain data should ideally not happen with the account key. During the development of the `ReleaseGold` smart contract it has been found that smart contracts as accounts cannot sign metadata. As a workaround, verifying metadata has been adjusted to allow any authorized signer to sign the payload, which violates the intention behind having well scoped authorized signers in the first place.
+- As part of ODIS, accounts have to authenticate with the ODIS signers for rate-limiting purposes. To do that, the "account key" or the "authorized DEK" has to sign the request which also violates the intention of keeping keys that can move funds as cold as possible.
+- For other applications such as CIP8 or other down-the-road applications, specific keys are going to be useful. Additionally, the `dataEncryptionKey` seems out of place without useful segmentation ability, despite it being effectively also an authorized key (if not signer).
+
+Allowing the segmentation of domains for a Celo account (i.e. an account on `Accounts.sol`) gives Celo stakeholders more flexibility in their custody setup. It allows sensitive domains (like moving balances) to be only accessible by keys that are "cold", i.e. harder to compromise, while making less sensitive domains more user friendly. Expanding such segmentation aids the user experience, both of interacting with the Celo Core Contracts themselves (via authorized smart contracts), as well as dapps that currently would need to copy this authorization scheme to receive similar benefits for their users. While this CIP does not specify further applications, per-dapp-authorizations can become practical.
+
+## Specification
+
+The proposed changes to `Accounts.sol`:
+
+- Add a `signerAuthorizations` struct:
+    ```
+    struct SignerAuthorization {
+      address signer
+      bool authorized
+    }
+    mapping(address => mapping(bytes32 => SignerAuthoriation)) public signerAuthorizations
+    ```
+- Add generalized `authorizeSigner(address signer, string calldata role)` and `completeSignerAuthorization(address account, string calldata role)` for smart contract authorizations (`authorizeSigner` creates a pending authorization that `completeSignerAuthorization` completes)
+- Add a generalized `authorizeSignerWithSignature(address signer, string calldata role, uint8 v, bytes32 r, bytes32 s)` (where v, r, s is an EIP712 signature) for proof-of-possession-based authorization (TODO: EIP712 domain specification)
+
+Both of the methods above would populate the relevant `SignerAuthorization` in the `signerAuthorizations` struct after hashing the `role`, as well as the `authorizedBy` struct. These methods should continue the invariant that a signer is only authorized by a single account ever. However, we can relax the invariant that a signer can only be authorized for at most 1 role.
+
+- Add a generalized `getAuthorizedSigner(address account, string calldata role)` for returning completed authorized signers.
```

Comment:
Currently, if a signer is not authorized the default is to [return the account address](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/common/Accounts.sol#L399-L408). What is the default for this updated system? Is there a way to have hierarchical keys such that if a dapp queries for `example.com/application/vote` and the key is not set but `example.com/application` is, it will return the more general key?

This could also be done by the caller by progressively querying up the path as long, provided that this function returns an indication that the key is not set. Providing native support for this would potentially make the lives of users easier, as there are already complaints about having to manage too many keys, so creating an easy way to establish a single "default signing key" and progressively additional keys as the user desires would be very helpful.

Reply:
I think that is really cool and sounds like something that can be done "entirely in user-land"? (though I certainly agree that codifying this here is more useful. Again, to keep the scope focused, do you think we can do this in a separate CIP?

Reply:
We can do that separately. It doesn't nessasrly need to be part of any CIP, and could instead just be a convention. The key question here is: What is returned when no signer is specified for a particular path?

Reply:
Conventions to me are like protocols, and should be encoded as CIPs where possible.

I believe the existing return when no signer is specified  is to return the account

Reply:
Alright. I think we should change that then. If the account is returned when no signer is specified, it will not be possible differentiate between when the signer is not specified. Instead I would recommend we return either the zero address and let the caller decide what to use as the default. (Since they already know the account address, it will be trivial for them to use it as the default, but this leaves the option for the caller to use another key path as a default instead.)

Reply:
That would break backwards compatibility though, right?

Reply:
I don't think so. This is a new API, so there shouldn't be any requirement to preserve the same exact behavior as the old one. If needed, the caller can default to the account address anyway.

Reply:
Good point for the new functions